File: pass1.py
class Assembler:
    def __init__(self):
        self.symbol_table = {}
        self.literal_table = {}
        self.intermediate_code = []
        self.location_counter = 0
        self.mot = self.create_mot()

    def create_mot(self):
        return {
            'ORIGIN': ('AD','03',1),
            'STOP': ('IS', '00',1),
            'ADD': ('IS', '01',1),
            'SUB': ('IS', '02',1),
            'MULT': ('IS', '03',1),
            'MOVER': ('IS', '04',1),
            'MOVEM': ('IS', '05',1),
            'COMP': ('IS', '06',1),
            'BC': ('IS', '07',1),
            'DIV': ('IS', '08',1),
            'READ': ('IS', '09',1),
            'PRINT': ('IS', '10',1),
            'START': ('AD', '01',1),
            'END': ('AD', '02',1),            
            'EQU': ('AD', '04',1),
            'LTORG': ('AD', '05',1),
            'DC': ('DL', '01',1),
            'DS': ('DL', '02',1),
        }

    def read_input(self, filename):
        with open(filename, 'r') as file:
            lines = file.readlines()
        return lines

    def parse_line(self, line):
        parts = line.split()
        label, opcode, operands = None, None, []

        if len(parts) == 4:
            label, opcode, operands = parts[0], parts[1], parts[2:]
        elif len(parts) == 3:
            if parts[1] in self.mot:
                label, opcode, operands = parts[0], parts[1], parts[2:]
            else:
                opcode, operands = parts[0], parts[1:]
        elif len(parts) == 2:
            if parts[0] in self.mot:
                opcode, operands = parts[0], parts[1:]
            else:
                label, opcode = parts[0], parts[1]
        elif len(parts) == 1:
            opcode = parts[0]

        return label, opcode, operands

    def process_line(self, label, opcode, operands):
        if label:
            if label not in self.symbol_table:
                self.symbol_table[label] = self.location_counter
            else:
                raise ValueError(f"Duplicate symbol: {label}")

        if opcode == 'START':
            self.location_counter = int(operands[0])
        elif opcode == 'END':
            self.process_literals()
            return
        elif opcode == 'LTORG':
            self.process_literals()
        else:
            if operands:
                for operand in operands:
                    if operand.startswith('='):
                        if operand not in self.literal_table:
                            self.literal_table[operand] = -1

            if opcode in self.mot:
                opcode_info = self.mot[opcode]
                self.intermediate_code.append((self.location_counter, label, opcode_info, operands))
                self.location_counter += opcode_info[2]
            else:
                self.intermediate_code.append((self.location_counter, label, opcode, operands))
                self.location_counter += 1

    def process_literals(self):
        for literal in list(self.literal_table.keys()):
            if self.literal_table[literal] == -1:
                self.literal_table[literal] = self.location_counter
                self.intermediate_code.append((self.location_counter, None, ('DL', '02', 1), literal))
                self.location_counter += 1

    def write_output(self, filename):
        with open(filename, 'w') as file:
            file.write("Intermediate Code:\n")
            for line in self.intermediate_code:
                file.write(f"{line}\n")
            file.write("\nSymbol Table:\n")
            for symbol, address in self.symbol_table.items():
                file.write(f"{symbol}: {address}\n")
            file.write("\nLiteral Table:\n")
            for literal, address in self.literal_table.items():
                file.write(f"{literal}: {address}\n")

    def assemble(self, input_file, output_file):
        lines = self.read_input(input_file)
        for line in lines:
            label, opcode, operands = self.parse_line(line.strip())
            self.process_line(label, opcode, operands)
        self.write_output(output_file)

if __name__ == "__main__":
    assembler = Assembler()
    assembler.assemble("/Users/yashraj146/Documents/LP1 Sem 5/Assember Python/input.asm",
                        "/Users/yashraj146/Documents/LP1 Sem 5/Assember Python/output.txt")

File: pass1.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <map>
#include <string>
#include <tuple>

using namespace std;

map<string, int> condition_codes = {
    {"LT", 1},
    {"LE", 2},
    {"EQ", 3},
    {"GT", 4},
    {"GE", 5},
    {"ANY", 6}
};

map<string, int> register_codes = {
    {"AREG", 1},
    {"BREG", 2},
    {"CREG", 3},
    {"DREG", 4}
};

map<string, tuple<string, int>> mnemonics = {
    {"STOP", {"IS", 0}},
    {"ADD", {"IS", 1}},
    {"SUB", {"IS", 2}},
    {"MULT", {"IS", 3}},
    {"MOVER", {"IS", 4}},
    {"MOVEM", {"IS", 5}},
    {"COMP", {"IS", 6}},
    {"BC", {"IS", 7}},
    {"DIV", {"IS", 8}},
    {"READ", {"IS", 9}},
    {"PRINT", {"IS", 10}},
    {"START", {"AD", 1}},
    {"END", {"AD", 2}},
    {"ORIGIN", {"AD", 3}},
    {"EQU", {"AD", 4}},
    {"LTORG", {"AD", 5}},
    {"DC", {"DL", 1}},
    {"DS", {"DL", 2}}
};

vector<pair<string, int>> symtab;

int find_symbol(const string& symbol_name) {
    for (size_t i = 0; i < symtab.size(); ++i) {
        if (symtab[i].first == symbol_name) return i;
    }
    return -1;
}

int update_symbol(const string& symbol_name, int addr) {
    for (size_t i = 0; i < symtab.size(); ++i) {
        if (symtab[i].first == symbol_name) {
            if (addr != -1) {
                symtab[i].second = addr;
            }
            return i;
        }
    }
    symtab.push_back(make_pair(symbol_name, addr));
    return symtab.size() - 1;
}

vector<pair<string, int>> lit_tab(10, make_pair("", -1));
vector<int> pool_tab(10, -1);
int lit_tab_ptr = 0;
int pool_tab_ptr = 0;
int location_cntr = 1;

void init_literals() {
    for (int i = pool_tab[pool_tab_ptr]; i < lit_tab_ptr; ++i) {
        lit_tab[i].second = location_cntr;
        ++location_cntr;
    }
    ++pool_tab_ptr;
    pool_tab[pool_tab_ptr] = lit_tab_ptr;
}

int main() {
    pool_tab[pool_tab_ptr] = 0;
    string source_file_path;
    cout << "Enter source code file path: ";
    cin >> source_file_path;

    ifstream file(source_file_path);
    if (!file.is_open()) {
        cerr << "Error: Could not open " << source_file_path << endl;
        return 1;
    }

    vector<string> source_lines;
    string line;
    while (getline(file, line)) {
        source_lines.push_back(line);
    }
    file.close();

    vector<vector<string>> source_line_tokens;
    for (const string& line : source_lines) {
        istringstream iss(line);
        vector<string> tokens;
        string token;
        while (iss >> token) {
            tokens.push_back(token);
        }
        source_line_tokens.push_back(tokens);
    }

    vector<string> ic_lines;

    for (const auto& line_tokens : source_line_tokens) {
        string label = "", mnemonic_str = "", operand1 = "", operand2 = "";

        if (mnemonics.count(line_tokens[0])) {
            mnemonic_str = line_tokens[0];
            if (line_tokens.size() == 2) operand1 = line_tokens[1];
            if (line_tokens.size() == 3) {
                operand1 = line_tokens[1];
                operand2 = line_tokens[2];
            }
        } else {
            label = line_tokens[0];
            mnemonic_str = line_tokens[1];
            if (line_tokens.size() == 3) operand1 = line_tokens[2];
            if (line_tokens.size() == 4) {
                operand1 = line_tokens[2];
                operand2 = line_tokens[3];
            }
        }

        string mnemonic_class = get<0>(mnemonics[mnemonic_str]);
        int mnemonic_opcode = get<1>(mnemonics[mnemonic_str]);
        string ic_line = (mnemonic_class != "AD" ? to_string(location_cntr) : "   ") + " (" + mnemonic_class + "," + to_string(mnemonic_opcode) + ") ";

        if (mnemonic_class != "DL" && !label.empty()) {
            update_symbol(label, location_cntr);
        }

        if (mnemonic_str == "LTORG") {
            init_literals();
        }

        if (mnemonic_str == "START") {
            location_cntr = stoi(operand1);
            ic_line += "(C," + to_string(location_cntr) + ")";
        }

        if (mnemonic_str == "ORIGIN") {
            size_t pos = operand1.find_first_of("+-");
            if (pos != string::npos) {
                string symbol_name = operand1.substr(0, pos);
                int symbol_index = find_symbol(symbol_name);
                int symbol_addr = symbol_index != -1 ? symtab[symbol_index].second : 0;
                location_cntr = stoi(operand1.replace(pos, 1, "") + to_string(symbol_addr));
                ic_line += "(C," + to_string(location_cntr) + ")";
            } else {
                int symbol_index = find_symbol(operand1);
                location_cntr = symbol_index == -1 ? stoi(operand1) : symtab[symbol_index].second;
                ic_line += "(C," + to_string(location_cntr) + ")";
            }
        }

        if (mnemonic_str == "EQU") {
            size_t pos = operand1.find_first_of("+-");
            int updated_val = 0;
            if (pos != string::npos) {
                string symbol_name = operand1.substr(0, pos);
                int symbol_index = find_symbol(symbol_name);
                int symbol_addr = symbol_index != -1 ? symtab[symbol_index].second : 0;
                updated_val = stoi(operand1.replace(pos, 1, "") + to_string(symbol_addr));
                update_symbol(label, updated_val);
                ic_line += "(S," + to_string(symbol_index) + ") (C," + to_string(updated_val) + ")";
            } else {
                int symbol_index = find_symbol(operand1);
                updated_val = symbol_index == -1 ? stoi(operand1) : symtab[symbol_index].second;
                update_symbol(label, updated_val);
            }
        }

        if (mnemonic_class == "DL") {
            int symbol_index = update_symbol(label, location_cntr);
            if (mnemonic_str == "DC") {
                location_cntr += 1;
                operand1 = operand1.substr(1, operand1.size() - 2);
            } else if (mnemonic_str == "DS") {
                location_cntr += stoi(operand1);
            }
            ic_line += "(S," + to_string(symbol_index) + ") (C," + operand1 + ")";
        }

        if (mnemonic_class == "IS") {
            if (mnemonic_str == "READ" || mnemonic_str == "PRINT") {
                int symbol_index = update_symbol(operand1, -1);
                ic_line += "(S," + to_string(symbol_index) + ")";
            } else if (mnemonic_str == "BC") {
                int symbol_index = update_symbol(operand2, -1);
                ic_line += "(" + to_string(condition_codes[operand1]) + ") (S," + to_string(symbol_index) + ")";
            } else if (mnemonic_str == "STOP") {
                // STOP has no operands
            } else {
                if (operand2.find('=') != string::npos) {
                    string literal_val = operand2.substr(operand2.find('=') + 1);
                    ic_line += "(" + to_string(register_codes[operand1]) + ") (L," + to_string(lit_tab_ptr) + ")";
                    lit_tab[lit_tab_ptr++] = make_pair(literal_val, -1);
                } else {
                    int symbol_index = update_symbol(operand2, -1);
                    ic_line += "(" + to_string(register_codes[operand1]) + ") (S," + to_string(symbol_index) + ")";
                }
            }
            location_cntr += 1;
        }

        ic_lines.push_back(ic_line);
    }

    init_literals();

    cout << "-------------- INTERMEDIATE CODE----------------" << endl;
    for (const auto& line : ic_lines) {
        cout << line << endl;
    }

    cout << "--------- SYMBOL TABLE -------------" << endl;
    for (const auto& entry : symtab) {
        cout << entry.first << " " << entry.second << endl;
    }

    cout << "--------- LITERAL TABLE -------------" << endl;
    for (int i = 0; i < lit_tab_ptr; ++i) {
        cout << lit_tab[i].first << " " << lit_tab[i].second << endl;
    }

    cout << "---------- POOL TABLE -------------" << endl;
    for (int i = 0; i < pool_tab_ptr; ++i) {
        cout << pool_tab[i] << endl;
    }

    return 0;
}